### PR TITLE
feat(studio): add system default option for theme

### LINKED ISSUE #4329: Add system default option for theme choice
# Chore

Currently only has dark and light:
image

see
https://stackoverflow.com/a/57795518

```
if (window.matchMedia && window.matchMedia('(prefers-color-scheme: dark)').matches) {
    // dark mode
}
```

Note that the theme choice selection should sync as well in the menu bar:
image

Behaviour:
- User preference page will have a third option to listen to system
- In the menu bar, the toggle should reflect if its currently dark or light
- toggling the theme in the menu bar will set it to dark or light (no longer system)

### PR DESCRIPTION
**Pull request by harshcut**
**Description**

Closes #4329. Enables _System Default_ theme option for Studio. Removed icon for `<Select.Option  />` when on `system` theme. The toggle inside the project updates with current system theme scheme. Updating the toggle overrides the current theme setting. Also added a check for bad theme string in `localStorage`. Added an event listener to update theme automatically when system scheme changes.

Edit: Changelog updated for commit 95f9e54d199489110ad5cb011deab20b1668466f.

### CODE REVIEW

File: studio/stores/UiStore.ts
Comment (phamhieu) [inline studio/stores/UiStore.ts:9]: Currently the studio app only supports `dark` and `light` theme. We don't have `system` theme. I believe there's a miss-understanding here.

Comment (phamhieu) [reply studio/stores/UiStore.ts:9]: According to https://github.com/supabase/supabase/issues/4329:
- adding a new `system` selection option on User preference page (DONE)
- checking window.localStorage.getItem('theme')
  - if `system`: Then we detect if the system is dark mode or not to set the theme value `dark` or `light`.
  - if `dark` or `light`: set theme value according to

**Review by phamhieu — CHANGES_REQUESTED** [review]

File: studio/stores/UiStore.ts
Comment (harshcut) [reply studio/stores/UiStore.ts:9]: Do you want me to check for `system` in `ui.load` method, and then accordingly set the theme to `light` or `dark`?

Comment (harshcut) [reply studio/stores/UiStore.ts:9]: - if `system`: Then we detect if the system is in dark mode or not to set the theme value `dark` or `light`.
- if `dark` or `light`: set theme value according to

I have already added a check for this in the [`ui.setTheme`](https://github.com/supabase/supabase/pull/4482/files#diff-39dba4c513e1bf9dcffbb0250c784c1b88d1543fcc8e97a074132bf39be31cddL88-R98) method. If you want `ui.setTheme()` to only have `light` or `dark` as props.

Comment (phamhieu) [reply studio/stores/UiStore.ts:9]: Let wait for @joshenlim confirmation. 

To me you only need to update load() method under UiStore to detect `theme` value. Your load method looks ok to me except 
```
load() {
  if (typeof window === "undefined") return;
  const savedTheme = (window.localStorage.getItem("theme") ?? "dark") as
    | "dark"
    | "light"
    | "system";
  if (savedTheme === "dark" || savedTheme === "light") {
   // why do you want to save theme value here again? 
    window.localStorage.setItem("theme", savedTheme);
    return this.setTheme(savedTheme);
  }
  if (savedTheme === "system") {
   // same question
    window.localStorage.setItem("theme", "system");
    return this.setTheme(
      window.matchMedia("(prefers-color-scheme: dark)").matches ? "dark" : "light"
    );
  }
  // same question
  window.localStorage.setItem("theme", "dark");
  this.setTheme("dark");
};
```

The rest `toggleTheme`, `setTheme `, `isDarkTheme ` should be kept as it is

Comment (phamhieu) [reply studio/stores/UiStore.ts:9]: Let clarify the term:
- theme-option: 3 dark, light and system
- theme: only 2 dark and light

Comment (harshcut) [reply studio/stores/UiStore.ts:9]: Sorry for that, I don't need to update the theme value in `localStorage` for every case. The value only needs to be updated if value received is not `light`, `dark` or `system`.

Comment (harshcut) [reply studio/stores/UiStore.ts:9]: @phamhieu If `ui.setTheme` method is suppose to have only `light` or `dark` as props, the method will be as follows. This is because the theme option in `localStorage` should be updated every time the `ui.seTheme()` is called outside the `ui.load`.

```ts
setTheme(theme: "dark" | "light", asSystem?: boolean) {
  if (asSystem) {
    this.theme = theme
    window.localStorage.setItem("theme", "system")
    if (theme === "light") return (document.body.className = "light")
    return (document.body.className = "dark")
  }
  this.theme = theme
  window.localStorage.setItem("theme", theme)
  document.body.className = theme
}
```

I still think using `system` as prop is more readable. This removes the need to fetch the `localStorage` every time we need to check if the theme option is set to `system` or not. It also removes the need for the `asSystem` prop. I will wait for the final decision and make the requested changes.

Comment (phamhieu) [reply studio/stores/UiStore.ts:9]: setTheme should not care about `theme-option` at all. 

When users select the system option, you can check the system config first then call UiStore.SetTheme with the final `theme` value.

Comment (harshcut) [reply studio/stores/UiStore.ts:9]: The theme options are `light`, `dark` and `system`. When the `system` theme-option is selected, we will pass the current system scheme (eg. `dark`) to the `setTheme` method and set the theme as `dark`, but this will also set the theme-option as `dark` and change the `localStorage` value to `dark`. So the next time the browser opens the theme will be inferred as `dark` and not `system`. Is this is what you mean?

Comment (phamhieu) [reply studio/stores/UiStore.ts:9]: > and change the localStorage value to dark

We save `theme-option` to local-storage NOT `theme`. And we already update `load()` method to support it.

Comment (phamhieu) [reply studio/stores/UiStore.ts:9]: I just check the current code for `setTheme`.  I'm so sorry for the confusion. Just to reconfirm
- we have to remove the save to localstorage inside `setTheme` and move it to the theme option selection page
- inside setTheme we 
https://github.com/supabase/supabase/blob/305fce8a019e89269c3d6152842c2ab5180e13c7/studio/stores/UiStore.ts#L88-L92

Comment (harshcut) [reply studio/stores/UiStore.ts:9]: So the only place where the theme-option will be stored is `localStorage`? And the `setTheme` method will not update the theme-option in `localStorage`, but will be done by functions that call `setTheme`?

**Comment by joshenlim** [comment]
hey @harshcut! awesome work here - sorry i only got to your PR now.
Thinking about it now properly tbh, I think let's just update the setting popup (via the left side nav) here to follow the profile preference page like this to a dropdown

image

that way i'd think it'll help to simplify some code logic here as well (remove the need to track a separate "option")

**Comment by harshcut** [comment]
image

**Review by phamhieu — APPROVED** [review]

File: studio/components/layouts/ProjectLayout/NavigationBar/NavigationBar.tsx
Comment (joshenlim) [inline studio/components/layouts/ProjectLayout/NavigationBar/NavigationBar.tsx:24]: @harshcut can we abstract these pieces of logic for the theme state, useEffect, and onThemeOptionChange into the UiStore?

The NavigationBar and  /account/me have the same pieces of logic so ideally we'd not want to repeat the code in both places. i.e the UiStore should be the source of truth, it should hold the `themeOption`, and NavigationBar and /account/me should just listen to that value instead of holding its own state

`onThemeOptionChange` can be placed within the UiStore as well and the `useEffect` can probably be placed in the `load` method of the UiStore

File: studio/stores/UiStore.ts
Comment (joshenlim) [inline studio/stores/UiStore.ts:86]: @harshcut probably good to double check this but I dont think we no longer use this `toggleTheme` method anymore now that we're moving away from the toggle in the NavigationBar. can remove this

**Review by joshenlim — CHANGES_REQUESTED** [review]
Just need to update shift some pieces of logic into the UiStore, instead of NavigationBar and /account/me having the same duplicated logic, but looking good! :)

File: studio/components/layouts/ProjectLayout/NavigationBar/NavigationBar.tsx
Comment (joshenlim) [inline studio/components/layouts/ProjectLayout/NavigationBar/NavigationBar.tsx:85]: value can listen to the UiStore directly, there's no need to keep a separate `value` state

i.e `<Select value={ui.themeOption} />`

File: studio/pages/account/me.tsx
Comment (joshenlim) [inline studio/pages/account/me.tsx:82]: likewise for here, no need to keep its own state, just listen directly to ui.themeOption

**Review by joshenlim — CHANGES_REQUESTED** [review]
just one small bit of detail left

File: studio/components/layouts/ProjectLayout/NavigationBar/NavigationBar.tsx
Comment (harshcut) [reply studio/components/layouts/ProjectLayout/NavigationBar/NavigationBar.tsx:85]: I already tried using that, but it did not update the value prop in `ThemeSettings`. I worked as expected in `NavigationBar`. What am I missing?

```tsx
function ThemeSettings() {
  const { ui } = useStore()

  return (
    <Panel
      title={[
        <Typography.Title key="panel-title" level={5}>
          Theme
        </Typography.Title>,
      ]}
    >
      <Panel.Content>
        <Select
          value={ui.themeOption}
          onChange={(e: any) => ui.onThemeOptionChange(e.target.value)}
        >
          <Select.Option value="system">System default</Select.Option>
          <Select.Option value="dark">Dark</Select.Option>
          <Select.Option value="light">Light</Select.Option>
        </Select>
      </Panel.Content>
    </Panel>
  )
}
```

**Review by joshenlim — APPROVED** [review]
Looking good! Awesome stuff @harshcut :) thanks so much! We'll look into getting your PR in after the holidays

**Pull request merged**
